src/mlpack/methods/lmnn/constraints_impl.hpp

```diff
@@ -145,33 +210,50 @@ void Constraints<MetricType>::Impostors(arma::Mat<size_t>& outputMatrix,
 // Calculates k differently labeled nearest neighbors. The function
 // writes back calculated neighbors & distances to passed matrices.
 template<typename MetricType>
-void Constraints<MetricType>::Impostors(arma::Mat<size_t>& outputNeighbors,
+void Constraints<MetricType>::Impostors(arma::Mat<size_t>& outputMatrix,
                                         arma::mat& outputDistance,
                                         const arma::mat& dataset,
-                                        const arma::Row<size_t>& labels)
+                                        const arma::Row<size_t>& labels,
+                                        const arma::mat& diffTransform)
 {
   // Perform pre-calculation. If neccesary.
   Precalculate(labels);
 
-  // KNN instance.
-  KNN knn;
-
   arma::Mat<size_t> neighbors;
   arma::mat distances;
 
   for (size_t i = 0; i < uniqueLabels.n_cols; i++)
   {
     // Perform KNN search with differently labeled points as reference
-    // set and  same class points as query set.
-    knn.Train(dataset.cols(indexDiff[i]));
-    knn.Search(dataset.cols(indexSame[i]), k, neighbors, distances);
+    // set and same class points as query set.
+    if (knnObjects.size() == uniqueLabels.n_elem)
+    {
+      auto Tree = knnObjects[i].ReferenceTree();
+
+      Tree.Dataset() = diffTransform * origDatasets[i];
+      UpdateTree(Tree);
+
+      knnObjects[i].Train(Tree);
```

Comment:
This line actually shouldn't be necessary---we've already updated the tree internally (assuming `UpdateTree()` is called on a reference).

Reply:
Done!
Sorry for not being available yesterday, I was shifting to my new dorm. It took some paperwork and a lot of hard work and almost took half of today's time :(

Reply:
No worries, I know how much work a move can be!  I hope the heavy lifting is all done at this point. :)